### ISSUE TITLE
Error: Provider produced inconsistent final plan - google_dns_record_set - rrdatas list with unknown value

### ISSUE THREAD

**Issue by zymotik** [hashicorp/terraform-provider-google#12043]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Issue:

Using a value that is unknown until apply in the rrdatas list of a google_dns_record_set resource causes an inconsistent plan error.

### Terraform Version

Terraform v1.2.4
on linux_amd64

### Affected Resource(s)

* google_dns_record_set

### Terraform Configuration Files

```tf
terraform {
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = ">= 4.24.0"
    }

    random = {
      version = ">= 3.3.2"
    }
  }
}

resource "random_id" "project_id" {
  keepers = {
    project_id = "my-project"
  }
  byte_length = 8
  prefix      = "project-"
}

resource "google_project" "my_project" {
  name = "my-${random_id.project_id.hex}"
  project_id = "${random_id.project_id.hex}"
}

resource "google_dns_record_set" "cdn_dns_txt_record_firebase" {
  name         = "www.example.com"
  project      = google_project.my_project.project_id
  managed_zone = "example.com."
  type         = "TXT"
  ttl          = 300
  rrdatas      = ["\"v=spf1 include:_spf.firebasemail.com ~all\"", "firebase=${google_project.my_project.id}"]
}
```

### Debug Output

https://gist.github.com/zymotik/5ad37a5534314b835d64a48358b8ddf6

### Panic Output

```
Error: Provider produced inconsistent final plan
When expanding the plan for google_dns_record_set.cdn_dns_txt_record_firebase to include new values learned so far during apply, provider
"registry.terraform.io/hashicorp/google" produced an invalid new value for .rrdatas: was null, but now cty.ListVal([]cty.Value{cty.StringVal("\"v=spf1
include:_spf.firebasemail.com ~all\""), cty.StringVal("firebase=projects/project-bf0a809a64d28d47")}).
 
This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

### Expected Behavior

Plan should include `rrdatas` element on the `google_dns_record_set.cdn_dns_txt_record_firebase` resource indicating value known after apply.

### Actual Behavior

Error reported:

Error: Provider produced inconsistent final plan

### Steps to Reproduce

1. `terraform apply`

### References

This issue maybe related:

* https://github.com/hashicorp/terraform-provider-google/issues/10131

**Comment by edwardmedia** [comment]
@zymotik do you plan to create the resource right after the project is created? It might take a certain time for the project to be ready. Can you try add sleep [logic](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/sleep) between two to see if that help solve the problem?

**Comment by zymotik** [comment]
An interesting suggestion, but I think the problem is that the value goes from null at plan to being a string on create. I can't see how delaying the resource at apply will fix the incorrectly computed null at the plan stage.

Am I misunderstanding how terraform plan works?

**Comment by edwardmedia** [comment]
@zymotik are you able to repro the issue if you separate these resources in different apply?

**Comment by zymotik** [comment]
> @zymotik are you able to repro the issue if you separate these resources in different apply? 

No, because if the Google Cloud project resource is already created in one apply, the next plan knows the values it needs in advance. The "random_id" resource does not generate a new id because of the "keepers" configuration.

The terraform example I have given is contrived. In the code we use, running it twice does work, but it's not an ideal solution.

**Comment by apparentlymart** [comment]
I think this item in the trace log points more directly to the cause of this oddity:

```
2022-07-07T13:10:47.261+0100 [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an invalid plan for google_dns_record_set.cdn_dns_txt_record_firebase, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .rrdatas: planned value cty.NullVal(cty.List(cty.String)) does not match config value cty.ListVal([]cty.Value{cty.StringVal("\"v=spf1 include:_spf.firebasemail.com ~all\""), cty.UnknownVal(cty.String)})
```

"because it is using the legacy plugin SDK", Terraform Core is reporting the problem as an inconsistency between the intermediate plan and the final plan, rather than between the configuration and the plan, but this warning shows that it's really the initial planning step that misbehaved here.

I wonder if this is a quirk of SDKv2's `DiffSuppressFunc` feature. We can see in the warning above that Terraform Core sees this value as a list containing one known element and one unknown element. I believe SDKv2 cannot support a partially-unknown list in this way, and so would typically simplify it as a wholly-unknown list.

However, by the time `DiffSuppressFunc` sees that value it has been serialized into the "flatmap" form that the SDK uses for the plan and state, so I think it would be shaped like this:

```go
map[string]string{
    // (...all of the other flatmap key/value pairs)
    "rrdatas.#": "74D93920-ED26-11E3-AC10-0800200C9A66",
}
```

That weird UUID above is the placeholder value that the SDK uses internally to represent "unknown value", defined internally as [`hcl2shim.UnknownVariableValue`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/internal/configs/hcl2shim#UnknownVariableValue).

If I'm right in all of this guesswork (which is far from guaranteed! :grinning:) then I think it _may_ be possible to address this by adding yet another exception to this early return branch at the top of the `DiffSuppressFunc`:

https://github.com/hashicorp/terraform-provider-google/blob/e31d0d84ad28cba926c43c9736342339b6620345/google/resource_dns_record_set.go#L16-L18

Something like this:

```go
 if k == "rrdatas.#" && (new == "0" || new == "74D93920-ED26-11E3-AC10-0800200C9A66") && old != new { 
 	return false 
 }
```

I don't know if the SDK exposes that `UnknownVariableValue` constant somewhere you can import from your provider -- the constant I linked to above is in an internal package and so not directly importable. If there _is_ a constant for it somewhere then I'd suggest using that rather than the raw value, to make this easier to understand in future. I see that [the Azure provider has its own copy of this constant](https://cs.github.com/hashicorp/terraform-provider-azuread/blob/7d917e39860c7200392e175f9e20ccd2ba1a4f98/internal/tf/pluginsdk.go?q=%2274D93920-ED26-11E3-AC10-0800200C9A66%22#L6) for a similar purpose, which is perhaps evidence that it's not exported anywhere reusable. :confounded: 

My intent with the above is that if the new value for this argument is unknown then it should _not_ "suppress the diff", and should therefore show the value as "known after apply" during the initial plan, rather than as null (omitted):

```
  # google_dns_record_set.cdn_dns_txt_record_firebase will be created
  + resource "google_dns_record_set" "cdn_dns_txt_record_firebase" {
      + id           = (known after apply)
      + managed_zone = "bob.com."
      + name         = "www.bob.com"
      + project      = (known after apply)
      + rrdatas      = (known after apply)
      + ttl          = 300
      + type         = "TXT"
    }
```

Then during the apply phase, Terraform Core will ask the provider for its final plan with `google_project.my_project.id` set to a known value, and so the provider will still get the opportunity to run its `DiffSuppressFunc` logic against that known value. Because the initial plan included an unknown value, Terraform will not _rule_ that the final "suppressed diff" is an "inconsistent final plan" (an unknown value can take on any value of the expected type) and so the apply step should complete as expected.

I want to be explicit that I've been working entirely in thought experiments here and have not actually tested any of this, because I don't have a GCP account to test with. If any of my assumptions don't hold in practice then the solution I proposed above probably won't actually fix it, but it seems worth a try!

**Comment by apparentlymart** [comment]
Followup note, for completeness: if I was right in my theory above, my proposed fix should avoid the error reported above but I think it will still show up as a warning message in the logs similar to the one I quoted above. The new one will be something like this:

```
2022-07-07T13:10:47.261+0100 [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an invalid plan for google_dns_record_set.cdn_dns_txt_record_firebase, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .rrdatas: planned value cty.UnknownVal(cty.List(cty.String)) does not match config value cty.ListVal([]cty.Value{cty.StringVal("\"v=spf1 include:_spf.firebasemail.com ~all\""), cty.UnknownVal(cty.String)})
```

Notice that _this_ error message says `planned value cty.UnknownVal(cty.List(cty.String))` instead of `planned value cty.NullVal(cty.List(cty.String))`.

_This_ inconsistency results from the fact that the legacy SDK cannot represent a partially-unknown list, and so this one is an example of the set of legacy SDK implementations that _require_ Terraform to treat this situation as a warning rather than an error. Terraform v0.11 and earlier didn't support partially-unknown lists and SDKv2 was written for Terraform v0.11 and earlier, so it produces answers that are valid for Terraform v0.11 but not for Terraform v0.12 or later; Terraform v0.12 or later "tolerate it" in order to keep all of these existing providers working even though they are written with the old SDK.

With that said then, this new warning is in some sense an "expected" warning: it's now describing a known limitation in the legacy SDK, rather than a bug in the provider. The original warning message was describing what is technically a bug in the provider logic, but Terraform doesn't have enough information to distinguish between SDK limitations and provider bugs at that stage so it just assumes it's an SDK limitation until it gets more information downstream, at which point it loudly classified it as a provider bug instead of a tolerable SDK limitation.

**Comment by edwardmedia** [comment]
@apparentlymart with below change, the issue is not resolved, with same error. Likely we need help from terraform core to resolve this. 

```
-       if k == "rrdatas.#" && new == "0" && old != new {
+       //if k == "rrdatas.#" && new == "0" && old != new {
+       if k == "rrdatas.#" && (new == "0" || new == "74D93920-ED26-11E3-AC10-0800200C9A66") && old != new { 
```

**Comment by apparentlymart** [comment]
@edwardmedia, I'm sorry my guesses were not correct!

That this does still seem to be an SDK-related issue or a provider-related issue, rather than Terraform Core issue, because we can see in the warning message that the `PlanResourceChange` response definitely did violate the expected rules, so this doesn't seem to be an example of Terraform Core misclassifying the response.

Given that, I did a little more poking in the relevant parts of the SDK, to try to understand what sort of input your provider code would likely be receiving in this situation where the list is partially unknown.

The first thing I noticed is that the SDK apparently emits a log message "Ignoring change due to DiffSuppressFunc" if the `DiffSuppressFunc` returns `true`, and now I know that I can confirm that this message does _not_ appear in the trace log and so it _seemed_ that we could disqualify my theory about the `DiffSuppressFunc` being the cause.

https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L1126-L1130

I did have one remaining uncertainty here about the `if !all` branch just above that log line. It isn't clear to me from this code what situations cause `all` to be `true` or `false`, but it is interesting to note both that it would generate no log line in that case _and_ it would cause the relevant key to be entirely missing from `diff.Attributes`, which could therefore make it appear as null in the final response to Terraform Core and therefore match the symptoms we can see. Unfortunately `m.diff` is called from _many_ different situations, and so I can't easily see just from code review what `all` would be set to in the case we're talking about here. Looking elsewhere in the file, I see the caller set `all` to `false` in the following situations:

https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L657
https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L684
https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L711
https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L725

We seem to set `all` to `true` only in the recursive call inside the `diffSet` function, which I suppose at least agrees with the comment in the original codepath I linked to above, which asserted that the special case relates to attributes of objects inside sets:

https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L1486
https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/schema.go#L1500

From the schema I can see that `rrdatas` is a list attribute and that it isn't nested inside a set, so I think we can safely assume that `all` will be `false` in the case we're worried about, which does unfortunately mean that the SDK would skip emitting the "Ignoring change due to DiffSuppressFunc" message, and so we actually _cannot_ elimitate the DiffSuppressFunc from consideration.

Given that, I find myself back where I started again, so I'm going to try to dig a bit deeper and try to answer the question of what a `DiffSuppressFunc` should expect to receive when the new value of its attribute is unknown. I'll leave another comment when I have more to share.

**Comment by apparentlymart** [comment]
Some more information! I'm sharing this just for completeness since I don't know yet what is useful, since I've not yet learned enough to actually have an answer to what's going on here.

The `d.GetChange` method wraps a hidden abstraction called a "field reader". The field reader for the _diff_ uses a shared helper function to deal with reading lists, which treats unknown lists and empty lists in the same way, but does signal unknown lists by populating the field "computed", which is some confusing legacy SDK terminology to mean "unknown":

https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/field_reader.go#L189-L196

Unfortunately, the `ResourceDiff.GetChange` method just discards everything except the value:

https://github.com/hashicorp/terraform-plugin-sdk/blob/4681738a561387fb0b3aaa69aeb42231383634a0/helper/schema/resource_data.go#L68-L77

So when you call `d.GetChange` there is no way to distinguish an unknown list from a known empty list.

**Comment by apparentlymart** [comment]
I think I've run out of ideas for what I can do just by code review and thought experiment, so I have some suggestions for next steps for gathering more information:

I would first suggest temporarily disabling the `DiffSuppressFunc` altogether to see if that makes the problem go away. Everything else I've been investigating here hinges on the `DiffSuppressFunc` somehow causing the problem, so if you can still reproduce the problem _without_ the `DiffSuppressFunc` then everything I've done so far and everything I'm about to suggest is moot, and it'll be time to search for an entirely new theory.

Assuming that removing `DiffSuppressFunc` _does_ appear to fix it, I would then suggest reinstating it and adding some extra trace logging to determine exactly what values you're getting as arguments, exactly what `d.GetChange` is returning, and what exactly `rrdatasListDiffSuppress` ends up doing when given that input.

My partial working theory at this time is that somehow this input is making the `DiffSuppressFunc` function return `true` when given this unknown list as input, which is then causing the SDK to omit that attribute entirely from its result. If that were true then the SDK's shims to the modern Terraform protocol would serialize that as a null list on the wire and thereby cause the warning I mentioned in my first comment. However, even if that is true I don't yet have a firm grasp on _why_ that `DiffSuppressFunc` function would be returning `true` with this input, because it looks like it should return `false` if there are zero items in the new value, which is what `d.GetChange` seems like it should be producing.

This is firmly in the murky depths of the SDK now, so if the debugging steps I described above don't lead to an answer immediately it'd probably be best to take the learnings here (both what I've found and described above, and the results of adding detailed trace logging to the function) to the SDK team and see what ideas they have.

Terraform Core doesn't "see" the distinction between the SDK and the provider here, so Terraform Core doesn't really care who makes this happen but the correct result would be to echo back the same partially-unknown list given in the configuration, and an _acceptable approximation_ of that would be to return a wholly-unknown list which Terraform Core will tolerate because this is the legacy SDK. I believe that once the provider and SDK taken together are _somehow_ returning one of those two acceptable outcomes this bug will be resolved, but it isn't yet clear to me whether it's the provider code or the SDK code that needs to change in order to make that happen.

**Comment by apparentlymart** [comment]
Hi @slevenick,

Yes, I think your result here indicates that. I believe `DiffSuppressFunc` in SDKv2 works by literally deleting all of the entries in the "diff" (an internal SDK structure) that are related to the attribute that the `DiffSuppressFunc` was attached to, and so if the _old_ value had no value for this attribute either then the SDK will end up constructing an object which has this attribute set to null, as we're seeing here. (The SDK constructs its planned value by merging the "diff" into the original value, so `DiffSuppressFunc` relies on there having been an old value that can shine through once the entries in the diff have been cleared out.)

From this output I think I can also see why what I suggested before didn't work: I was expecting "new" to be the "unknown value" placeholder string `"74D93920-ED26-11E3-AC10-0800200C9A66"`, but it seems like the SDK is trying to hide that implementation detail from you by setting `New` to be an empty string instead:

```
[DEBUG] Old is "0": timestamp=2022-07-20T18:15:43.482-0700
[DEBUG] New is "": timestamp=2022-07-20T18:15:43.482-0700
```

Given that, I wonder if modifying what I suggested earlier to use the empty string instead would help:

```go
 if k == "rrdatas.#" && (new == "0" || new == "") && old != new { 
 	return false 
 }
```

Just from observing the behavior from the outside (I've not confirmed this in code) it _seems_ like the SDK differentiates between "there are definitely zero elements" and "we don't know how many elements there are" by passing `"0"` for the first case and `""` for the second case. But I guess that distinction doesn't really matter here either way, since the `DiffSuppressFunc`'s treatment of both will be the same anyway.

---

If this _does_ work out then this seems like something worth documenting in the SDKv2 documentation for `DiffSuppressFunc`. I personally find it surprising that the SDK calls `DiffSuppressFunc` _at all_ when it's planning a "create" action, because its supposed purpose is to recognize when a new value is functionally equivalent to an old value and so that really only makes sense if there _is_ an "old value". But I expect like most things in SDKv2 it's not changeable without unwanted consequences to existing providers, and so documenting some different ways for a `DiffSuppressFunc` to notice it's being called during the create phase and just immediately `return false` if so.

`schema.ResourceData` has a method [`IsNewResource`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk@v1.17.2/helper/schema#ResourceData.IsNewResource) which I believe is intended to return `true` if the SDK thinks it's planning a create, but I notice that `schema.ResourceDiff` doesn't have a matching method and so that information isn't available to `DiffSuppressFunc`. Maybe the SDK team would consider adding it there so that we could document a totally-boilerplate solution that would work for any `DiffSuppressFunc`:

```go
if d.IsNewResource() {
 	return false
}
```

I'm getting ahead of myself, though. If the local fix I described above works then I'd suggest moving forward with that for now, but might also be worth taking this discussion upstream into [the SDKv2 repository](https://github.com/hashicorp/terraform-plugin-sdk) so we can get the SDK team's take on possible solutions that might help in other cases like this one.
